crates/bevy_ecs/src/bundle.rs

```diff
@@ -75,26 +77,107 @@ use std::{any::TypeId, collections::HashMap};
 /// bundle, in the _exact_ order that [`Bundle::get_components`] is called.
 /// - [`Bundle::from_components`] must call `func` exactly once for each [`ComponentId`] returned by
 ///   [`Bundle::component_ids`].
-pub unsafe trait Bundle: Send + Sync + 'static {
-    /// Gets this [`Bundle`]'s component ids, in the order of this bundle's [`Component`]s
-    fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId>;
-
+pub unsafe trait Bundle: ApplicableBundle + Sized {
     /// Calls `func`, which should return data for each component in the bundle, in the order of
     /// this bundle's [`Component`]s
     ///
     /// # Safety
     /// Caller must return data for each component in the bundle, in the order of this bundle's
     /// [`Component`]s
-    unsafe fn from_components(func: impl FnMut() -> *mut u8) -> Self
-    where
-        Self: Sized;
+    unsafe fn from_components(func: impl FnMut() -> *mut u8) -> Self;
+
+    /// Gets this [`Bundle`]'s component ids, in the order of this bundle's [`Component`]s
+    fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId>;
 
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This will
     /// [`std::mem::forget`] the bundle fields, so callers are responsible for dropping the fields
     /// if that is desirable.
     fn get_components(self, func: impl FnMut(*mut u8));
 }
 
+mod sealed {
+    use super::{ApplicableBundle, Bundle};
+
+    pub trait SealedApplicableBundle {}
+    impl<T> SealedApplicableBundle for T where T: Bundle {}
+    impl SealedApplicableBundle for Box<dyn ApplicableBundle> {}
+}
+
+/// A bundle which can be applied to a world
+///
+/// # Safety:
+/// The bundle returned from `init_bundle_info` is the same as used for `get_component_box`.
+///
+/// This trait is sealed and cannot be implemented outside of `bevy_ecs`
+///
+/// However, it is implemented for every type which implements [`Bundle`]
+
+pub unsafe trait ApplicableBundle:
```

Comment:
Maybe call this `DynBundle`?

Reply:
Possibly? It is not limited to dynamic use cases; just cases where `self` is available. This is the main entry way to 'single bundle insertion'.

Normal `Bundle` is absolutely static, but that doesn't mean that `ApplicableBundle` is not static.

Reply:
My preference here is either:

1. Call this `Bundle`, and use `StaticBundle` for the other trait.
2. Call this `InsertableBundle` or `CreationBundle`.

The former is nice, because in 99% of cases, users only need the methods from this trait to work with what they need. Removing bundles is typically not very useful. However it's a bit confusing, as deriving Bundle would derive a trait and its super trait.
 
The latter is just playing with synonyms to better reflect intent in terms of things users are familiar with.

Reply:
I did consider option 1. My concern was that it requires any manual impls of `Bundle` to know that it's actually a different trait which gets implemented. This change is currently non-breaking (as far as I can tell), which is quite neat.

Reply:
IMO I think that manual impls of `Bundle` are going to be exceedingly rare, and the compiler errors should be quite useful. The fact that `Bundle` would now be unsafe would also encourage people to Read the Docs.

Reply:
? `Bundle` has always been `unsafe`. I do agree though that manual impls are going to be rare.

Reply:
One other thing is that the renaming would require a careful audit of documentation etc. as `Bundle` can no longer be used for removal operations. I don't really want to do that in this PR if it can be avoided.